doc/TockBinaryFormat.md

```diff
@@ -0,0 +1,127 @@
+# Tock Binary Format
+
+<!-- npm i -g markdown-toc; markdown-toc -i Compilation.md -->
+
+<!-- toc -->
+
+- [TBF Header](#tbf-header)
+  * [TLV Types](#tlv-types)
+    + [`1` Main](#1-main)
+  * [`2` Writeable Flash Region](#2-writeable-flash-region)
+  * [`3` Package Name](#3-package-name)
+- [Code](#code)
+
+<!-- tocstop -->
+
+Tock processes are represented using the Tock Binary Format (TBF). A TBF
+includes a header portion, which encodes meta-data about the process, followed
+by a binary blob which is executed directly.
+
+## TBF Header 
+
+The TBF header contains a base header, followed by a sequence of
+type-length-value encoded elements. The base header 16 bytes, and has 5 fields:
+
+```
+ 0         2             4            8       12         16
++---------+-------------+------------+-------+----------+
+| Version | Header Size | Total Size | Flags | Checksum |
++---------+-------------+------------+-------+----------+
+```
+
+    * `Version` a 16-bit unsigned integer specifying the TBF header version.
+      Always `2`.
+    * `Header Size` a 16-bit unsigned integer specifying the length of the TBF
+      header in bytes.
+    * `Total Size` a 32-bit unsigned integer specifying the total size of the TBF
+      in bytes (including the header).
+    * `Flags` each bit indicates whether a flag is enabled (1) or disabled (0).
+      - Bit 0 marks the process enabled. Disabed processes will not be launched
+        at startup.
+    * `Checksum` the result of XORing each 4-byte word in the header, excluding
+      the word containing the checksum field itself.
+
+The header is followed immediately by a sequence of TLV elememnts.  Each
+element begins with a 16-bit type and 16-bit length followed by the element
+data:
```

Comment:
Need to note that all TLV headers must be aligned to a multiple of 4 bytes and that there can be up to 3 bytes of padding after the data section of a TLV.